sgkit_plink/pysnptools.py

```diff
@@ -0,0 +1,252 @@
+        Delimiter for `.bim` file, by default " "
+    bim_int_contig : bool, optional
+        Whether or not the contig/chromosome name in the `.bim`
+        file should be interpreted as an integer, by default False.
+        If False, then the `variant/contig` field in the resulting
+        dataset will contain the indexes of corresponding strings
+        encountered in the first `.bim` field.
+    count_a1 : bool, optional
+        Whether or not allele counts should be for A1 or A2,
+        by default True. Typically A1 is the minor allele
+        and should be counted instead of A2.  This is not enforced
+        by PLINK though and it is up to the data generating process
+        to ensure that A1 is in fact an alternate/minor/effect
+        allele. See https://www.cog-genomics.org/plink/1.9/formats
+        for more details.
+    lock : bool, optional
+        Whether or not to synchronize concurrent reads of `.bed`
+        file blocks, by default False. This is passed through to
+        [dask.array.from_array](https://docs.dask.org/en/latest/array-api.html#dask.array.from_array).
+    persist : bool, optional
+        Whether or not to persist `.fam` and `.bim` information in
+        memory, by default True.  This is an important performance
+        consideration as the plain text files for this data will
+        be read multiple times when False. This can lead to load
+        times that are upwards of 10x slower.
+
+    Returns
+    -------
+    Dataset
+        A dataset containing genotypes as 3D calls along with
+        all accompanying pedigree and variant information.
+    """
+
+    # Load axis data first to determine dimension sizes
+    df_fam = read_fam(path, sep=fam_sep)
+    df_bim = read_bim(path, sep=bim_sep)
+
+    # TODO: not on nearly 10x faster on real datasets
+    if persist:
+        df_fam = df_fam.persist()
+        df_bim = df_bim.persist()
+
+    arr_fam = _to_dict(df_fam, dtype=FAM_ARRAY_DTYPE)
+    arr_bim = _to_dict(df_bim, dtype=BIM_ARRAY_DTYPE)
+
+    # Load genotyping data
+    call_genotype = da.from_array(
+        # Make sure to use asarray=False in order for masked arrays to propagate
+        BedReader(path, (len(df_bim), len(df_fam)), count_A1=count_a1),
+        chunks=chunks,
+        # Lock must be true with multiprocessing dask scheduler
+        # to not get pysnptools errors (it works w/ threading backend though)
+        lock=lock,
+        asarray=False,
+        name=f"pysnptools:read_plink:{path}",
+    )
+
+    # If contigs are already integers, use them as-is
+    if bim_int_contig:
+        variant_contig = arr_bim["contig"].astype("int16")
+        variant_contig_names = da.unique(variant_contig).astype(str)
+        variant_contig_names = list(variant_contig_names.compute())
+    # Otherwise index by unique name where index will correspond
+    # to lexsort on names, i.e. if contigs are 'chr1', 'chr2',
+    # ..., 'chr10' then 'chr10' comes before 'chr2'
```

Comment:
I was thinking about that when I saw your comment in your initial changes.  Personally, I thought it would make more sense to say that using the contig name index is a better practice (we'll need to support that well anyhow).  Referring to them directly by index when not already encoded as ints seemed like it would invite errors.  In other words, I thought this was reasonable:

```
ds.sel(variant=ds.contig == ds.attrs['contigs'].index('chr1'))
```

I don't feel strongly about it though if you think access by order-of-appearance index will be important.

Reply:
I've seen problems with sorting contig names in the past (in BAM/VCF headers), so I thought ordering by appearance would be making slightly fewer assumptions. (There's a neat way of doing it here BTW: https://stackoverflow.com/a/12926989).

Doesn't your `ds.sel` work whatever the order of `contigs` - i.e. lex sorted or oder-of-appearance sorted?

Reply:
Yea the `ds.sel` using `.attrs['contigs']` should work regardless so I think we should push ourselves and users towards it.  

One advantage of the order-of-appearance indexing though would be that contig indexes in an Xarray index wouldn't cause a re-ordering relative to the original dataset.  I can't think of any access patterns where sequential access across contigs is important (e.g. [range queries](https://discourse.smadstatgen.org/t/n-d-array-conventions-for-variation-data/16/2?u=eczech)), so the order of them shouldn't matter, but reordering contigs should lead to Dask needlessly running big slices on the other non-indexed arrays.

I'll change it but if the contig index isn't going to be an internal detail, it makes me wonder if we should use 1-based indexing.  I think most geneticists would find 0-based contig codes problematic.

Reply:
It's using Python indexing, so 0-based makes sense I think?

Agreed on the other points.

Reply:
Changed in https://github.com/pystatgen/sgkit-plink/pull/1/commits/717d307dab6f4bd35088cb3896ca84d65c312256

Reply:
> ds.sel(variant=ds.contig == ds.attrs['contigs'].index('chr1'))

Btw I'm cool with this but it will be a bit obscure for new users. Is there any way we could make some convenience or shortcut?

Reply:
A few options that come to mind:

- Using an accessor:
```python 
ds.genetics.sel_contig(slice('chr1'))
``` 
- Accessor w/ attribute specific function:
```python
ds.sel(variant=ds['variant/contig'] == ds.meta.contig_index('chr1'))
``` 
- Use contig index directly:
```python
ds.sel(variant=ds['variant/contig'] == 0) # for chr1
```

I think it makes more sense to focus on the situation with an index though, which is a bit more clear on it's own IMO:

```python
contig_index = ds.attrs['contigs'].index('chr1')
contig_range = slice(contig_index, contig_index)
# or contig_range = slice(0, 0) for chr1
bp_pos_range = slice(0, 1000000)
ds.set_index(variant=('variant/contig', 'variant/pos')).sel(variant=(contig_range, bp_pos_range))
```

Any thoughts/preferences on those?

Reply:
Thanks @eric-czech for thinking through options. Moved this discussion over to https://github.com/pystatgen/sgkit/issues/25 to give it some more space.